redash/utils/sql_query.py

```diff
@@ -0,0 +1,45 @@
+import sqlparse
+from redash.utils import mustache_render
+from cached_property import cached_property
+
+
+class SQLQuery(object):
+    def __init__(self, template):
+        self.template = template
+        self.query = template
+
+    def apply(self, parameters):
+        if self.__dict__.has_key('text'):
+            del self.__dict__['text']
+
+        self.query = mustache_render(self.template, parameters)
+        return self
+
+    def is_safe(self):
```

Comment:
`is_safe` should perform a few additional tests before doing the SQL tree comparison:

1. Test that the passed value is of the type the parameter expects (in case the parameter type isn't a string). 
2. If it's a dropdown (query based or static) parameter, then compare with the list of allowed values.

Reply:
I tend to think that this should be in a different module. These checks are more related to the validity and compatibility of the provided parameters, and that's a different type of knowledge. I'd prefer to address this outside the current `SQLQuery` class (and maybe tree validations should be moved out as well, and `SQLQuery` should integrate these two distinct responsibilities - that's something that should be considered outside the scope of this PR, IMO)

Reply:
The implementation of this logic should be outside this class, as it will be reused in "JSONQuery" and others. Whether this class should call it some outside caller, is probably something we can decide when we get to integrate it with the actual codebase.